pkg/controller/node/nodecontroller.go

```diff
@@ -482,37 +487,97 @@ func (nc *NodeController) monitorNodeStatus() error {
+		if _, had := nc.zoneStates[k]; !had {
+			nc.zoneStates[k] = stateInitial
 		}
-		for i := range nodes.Items {
-			if utilnode.GetZoneKey(&nodes.Items[i]) == k {
-				if newState == stateFullSegmentation {
-					// When zone is fully segmented we stop the eviction all together.
-					nc.cancelPodEviction(&nodes.Items[i])
-				}
-				if newState == stateNormal && nc.zoneStates[k] == stateFullSegmentation {
-					// When exiting segmentation mode update probe timestamps on all Nodes.
-					now := nc.now()
-					v := nc.nodeStatusMap[nodes.Items[i].Name]
-					v.probeTimestamp = now
-					v.readyTransitionTimestamp = now
-					nc.nodeStatusMap[nodes.Items[i].Name] = v
-				}
-			}
+	}
+
+	allWasFullyDisrupted := true
+	for _, v := range nc.zoneStates {
+		if v != stateFullDisruption {
+			allWasFullyDisrupted = false
+			break
 		}
-		nc.zoneStates[k] = newState
 	}
 
-	return nil
+	// At least one node was responding in previous pass or in the current pass. Semantics is as follows:
+	// - if the new state is "partialDisruption" we call a user defined function that returns a new limiter to use,
+	// - if the new state is "normal" we resume normal operation (go back to default limiter settings),
+	// - if new state is "fullDisruption" we restore normal eviction rate,
+	//   - unless all zones in the cluster are in "fullDisruption" - in that case we state we stop all evictions.
+	if !allAreFullyDisrupted || !allWasFullyDisrupted {
+		// We're switching to full disruption mode
+		if allAreFullyDisrupted {
+			glog.V(0).Info("NodeController detected that all Nodes are not-Ready. Entering master disruption mode.")
+			for i := range nodes.Items {
+				nc.cancelPodEviction(&nodes.Items[i])
+			}
+			// We stop all evictions.
+			for k := range nc.zonePodEvictor {
+				nc.zonePodEvictor[k].SwapLimiter(0)
+				nc.zoneTerminationEvictor[k].SwapLimiter(0)
+			}
+			for k := range nc.zoneStates {
+				nc.zoneStates[k] = stateFullDisruption
+			}
+			// All rate limiters are updated, so we can return early here.
+			return
+		}
+		// We're exiting full disruption mode
+		if allWasFullyDisrupted {
+			glog.V(0).Info("NodeController detected that some Nodes are Ready. Exiting master disruption mode.")
+			// When exiting disruption mode update probe timestamps on all Nodes.
+			for i := range nodes.Items {
+				now := nc.now()
+				v := nc.nodeStatusMap[nodes.Items[i].Name]
+				v.probeTimestamp = now
+				v.readyTransitionTimestamp = now
+				nc.nodeStatusMap[nodes.Items[i].Name] = v
+			}
+			// We reset all rate limiters to default. Because we reset all timers we have plenty of time to slow them down
+			// in some nodes are now partially disrupted.
+			for k := range nc.zonePodEvictor {
+				nc.zonePodEvictor[k].SwapLimiter(
+					nc.enterPartialDisruptionFunc(len(zoneToNodeConditions[k]), nc.evictionLimiterQPS))
+				nc.zoneTerminationEvictor[k].SwapLimiter(
+					nc.enterPartialDisruptionFunc(len(zoneToNodeConditions[k]), nc.evictionLimiterQPS))
```

Comment:
The comment says that we reset rate-limitters to default, but you are changing the to "partial disruption" which is reduced (not default). Something is wrong here...